Implement errorOccured in webNavigation.getFrame/getAllFrames

b-weinstein opened the pull request:
#### baf3eae87da202e1ab9d7ccb2dff883fe503e2bc
<pre>
Implement errorOccured in webNavigation.getFrame/getAllFrames
<a href="https://bugs.webkit.org/show_bug.cgi?id=266499">https://bugs.webkit.org/show_bug.cgi?id=266499</a>
<a href="https://rdar.apple.com/118340990">rdar://118340990</a>

Reviewed by Alex Christensen and Timothy Hatcher.

Before this change, there was no way to determine if an error occurred in the load of a frame given a WKFrameInfo.

This patch pipes that information from the FrameLoader -&gt; WebFrame -&gt; WKFrameInfo, and exposes it as SPI on the WKFrameInfo.

This patch also adds some tests for both the new WKFrameInfo SPI and how it is exposed to extensions.

* Source/WebCore/loader/FrameLoader.cpp: Clear m_errorOccurredInLoading in all of the places that load start.
(WebCore::FrameLoader::dispatchDidFailProvisionalLoad): Set m_errorOccurredInLoading to true.
(WebCore::FrameLoader::checkLoadCompleteForThisFrame): Set m_errorOccurredInLoading to true if necessary.
* Source/WebCore/loader/FrameLoader.h:
* Source/WebKit/Shared/FrameInfoData.h: Add the new parameter.
* Source/WebKit/Shared/FrameInfoData.serialization.in: Ditto.
* Source/WebKit/UIProcess/API/APIFrameInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/WKFrameInfo.mm:
(-[WKFrameInfo _errorOccurred]):
* Source/WebKit/UIProcess/API/Cocoa/WKFrameInfoPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm:
(WebKit::frameParametersForFrame): Include whether or not an error occurred.
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::info const): Include whether or not an error occurred.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebNavigation.mm:
(TestWebKitAPI::TEST): Add a test that when a frame load fails with an error, errorOccurred is set in the getFrame call. There
are two flavors of this test, one with a failed provisional load, the other with a failed load after it had been committed.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm: Enhance the Frames test to:
1) Make sure the WKFrameInfo _errorOccurred SPI value is correct when loads fail and complete.
2) Perform an additional load after the failed load to make sure the error occurred state is reset.

Canonical link: <a href="https://commits.webkit.org/272303@main">https://commits.webkit.org/272303@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9be4906a66bac5c1f6ebc2b3c84ef9b1dce1809b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31179 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9857 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32873 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33687 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28318 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/31946 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12199 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7115 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27968 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31516 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8310 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27875 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7138 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7315 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27770 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35029 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28377 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28226 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33451 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7353 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5410 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31290 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9038 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8054 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4071 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7887 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm:
```diff
@@ -51,8 +51,7 @@ static WebExtensionFrameParameters frameParametersForFrame(_WKFrameTreeNode *fra
         // errorOccured
-        // FIXME: Correctly populate this based on whether or not an error occurred loading this frame
-        false,
+        (bool)frameInfo._errorOccurred,
```
The cast was required?

b-weinstein replied:
Yeah - `BOOL` vs `bool`.

xeenon replied:
Bummer.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebNavigation.mm:
```diff
@@ -421,6 +421,56 @@
+        // Setup
+        @"function errorListener(details) {",
+        // A subframe should have been the one to have the error.
+        @"  browser.test.assertFalse(details.frameId == 0)",
```
`browser.test.assertEq(details.frameId, 0)`

b-weinstein replied:
This is `assertFalse` - I want them to _not_ be equal.

b-weinstein replied:
Is there an `assertNeq`?

xeenon replied:
Oh! Then `assertFalse` is fine.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebNavigation.mm:
```diff
@@ -421,6 +421,56 @@
+        @"  browser.webNavigation.getFrame({ tabId: details.tabId, frameId: details.frameId }, function(frame) {",
+        @"    browser.test.assertEq(frame.parentFrameId, 0)",
+        @"    browser.test.assertTrue(frame.errorOccurred)",
+        // One thing to note here is that if the provisional load fails, there won't be a URL in the details.
```
You can assert this:

```
browser.test.assertEq(typeof frame.url, 'undefined')
```

Or should it be empty string? Either way you should test for it.

b-weinstein replied:
I believe it would be the empty string. I can add a test.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1141,6 +1141,8 @@ void FrameLoader::provisionalLoadStarted()
             page->didStartProvisionalLoad();
     }
+
+    m_errorOccurredInLoading = false;
```
This seems like it's in the wrong place, and it needs a test that verifies a successful load after an unsuccessful load indicates no more load failure.

b-weinstein replied:
As I mentioned in the other comment - this is for the provisional loading case and I have another place I set it in the committed load case.

How would you recommend writing such a test? How should I get the same frame to load a failing URL and then a valid URL?

b-weinstein replied:
Oh - I misread this line when saying it's in the wrong place. Where do you think it should be when a brand new load starts @achristensen07?

b-weinstein replied:
After looking more closely, I would guess `FrameLoader::load` is a better place.

b-weinstein replied:
Although there are lots of those flavors - I could put it in all of them.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2570,6 +2572,7 @@ CachePolicy FrameLoader::subresourceCachePolicy(const URL& url) const
 void FrameLoader::dispatchDidFailProvisionalLoad(DocumentLoader& provisionalDocumentLoader, const ResourceError& error, WillInternallyHandleFailure willInternallyHandleFailure)
 {
     m_provisionalLoadErrorBeingHandledURL = provisionalDocumentLoader.url();
+    m_errorOccurredInLoading = true;
```
This seems to only handle provisional load failures, which is the only thing tested.  We need to also cover and test the case where a load fails after it has committed.

TEST(WKDownload, FailAndResume) has such a failure.  It has the server respond with Content-Length: 10000 then sends 5k then terminates.  It may more reliably commit if it uses a content-length of 10mb and sends 5mb then terminates.

b-weinstein replied:
`m_errorOccurredInLoading = true` is also set in `FrameLoader::checkLoadCompleteForThisFrame`

```
        auto& error = documentLoader->mainDocumentError();

        AXObjectCache::AXLoadingEvent loadingEvent;
        if (!error.isNull()) {
            FRAMELOADER_RELEASE_LOG(ResourceLoading, "checkLoadCompleteForThisFrame: Finished frame load with error (isTimeout = %d, isCancellation = %d, errorCode = %d)", error.isTimeout(), error.isCancellation(), error.errorCode());
            m_client->dispatchDidFailLoad(error);
            loadingEvent = AXObjectCache::AXLoadingFailed;
            m_errorOccurredInLoading = true;
        } else {
            FRAMELOADER_RELEASE_LOG(ResourceLoading, "checkLoadCompleteForThisFrame: Finished frame load");
#if ENABLE(DATA_DETECTION)
            RefPtr document = m_frame->document();
            OptionSet<DataDetectorType> types = m_frame->settings().dataDetectorTypes();
            if (document && types) {
                auto documentLevelResults = retainPtr(DataDetection::detectContentInRange(makeRangeSelectingNodeContents(*document), types, m_client->dataDetectionReferenceDate()));
                m_frame->dataDetectionResults().setDocumentLevelResults(documentLevelResults.get());
                if (m_frame->isMainFrame())
                    m_client->dispatchDidFinishDataDetection(documentLevelResults.get());
            }
#endif
            m_client->dispatchDidFinishLoad();
            loadingEvent = AXObjectCache::AXLoadingFinished;
        }
```

Right before dispatchDidFailLoad is called. I believe that would cover the case where the load fails after it has committed.

I do agree that I need a test for that though.

b-weinstein replied:
I added the test for this.

b-weinstein committed:
Implement errorOccured in webNavigation.getFrame/getAllFrames
https://bugs.webkit.org/show_bug.cgi?id=266499
rdar://118340990

Reviewed by Alex Christensen and Timothy Hatcher.

Before this change, there was no way to determine if an error occurred in the load of a frame given a WKFrameInfo.

This patch pipes that information from the FrameLoader -> WebFrame -> WKFrameInfo, and exposes it as SPI on the WKFrameInfo.

This patch also adds some tests for both the new WKFrameInfo SPI and how it is exposed to extensions.

* Source/WebCore/loader/FrameLoader.cpp: Clear m_errorOccurredInLoading in all of the places that load start.
(WebCore::FrameLoader::dispatchDidFailProvisionalLoad): Set m_errorOccurredInLoading to true.
(WebCore::FrameLoader::checkLoadCompleteForThisFrame): Set m_errorOccurredInLoading to true if necessary.
* Source/WebCore/loader/FrameLoader.h:
* Source/WebKit/Shared/FrameInfoData.h: Add the new parameter.
* Source/WebKit/Shared/FrameInfoData.serialization.in: Ditto.
* Source/WebKit/UIProcess/API/APIFrameInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/WKFrameInfo.mm:
(-[WKFrameInfo _errorOccurred]):
* Source/WebKit/UIProcess/API/Cocoa/WKFrameInfoPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm:
(WebKit::frameParametersForFrame): Include whether or not an error occurred.
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::info const): Include whether or not an error occurred.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebNavigation.mm:
(TestWebKitAPI::TEST): Add a test that when a frame load fails with an error, errorOccurred is set in the getFrame call. There
are two flavors of this test, one with a failed provisional load, the other with a failed load after it had been committed.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm: Enhance the Frames test to:
1) Make sure the WKFrameInfo _errorOccurred SPI value is correct when loads fail and complete.
2) Perform an additional load after the failed load to make sure the error occurred state is reset.

Canonical link: https://commits.webkit.org/272303@main
